apps/labs/posts/scipy-array-api.md

```diff
@@ -0,0 +1,441 @@
+these assertions.
+The assertions make use of functions from `xp` where possible which helps avoid device transfers.
+
+Other than that, we make similar changes to those in the production functions so that every test works
+with all of the backends we test with.
+Finally, for tests of functions which hit compiled code, we add the `@skip_if_array_api_gpu` decorator.
+
+Here is what that looks like for one of the Discrete Sine and Cosine Transforms tests:
+
+```diff
++ @skip_if_array_api_gpu
++ @array_api_compatible
+  @pytest.mark.parametrize("func", ['dct', 'dst', 'dctn', 'dstn'])
+  @pytest.mark.parametrize("type", [1, 2, 3, 4])
+  @pytest.mark.parametrize("norm", [None, 'backward', 'ortho', 'forward'])
+- def test_fftpack_equivalience(func, type, norm):
++ def test_fftpack_equivalience(func, type, norm, xp):
+      x = np.random.rand(8, 16)
++     fftpack_res = xp.asarray(getattr(fftpack, func)(x, type, norm=norm))
++     x = xp.asarray(x)
+      fft_res = getattr(fft, func)(x, type, norm=norm)
+-     fftpack_res = getattr(fftpack, func)(x, type, norm=norm)
+
+-     assert_allclose(fft_res, fftpack_res)
++     xp_assert_close(fft_res, fftpack_res)
+```
+
+## Machinery, Helpers and Utilities
+
+I'll talk a bit about implementation details now - feel free to skip this section if you are just here for
+an overview!
+
+When we call `array_namespace`, a little more is done than just returning the namespace.
+If the environment variable `SCIPY_ARRAY_API` is not set, `array_api_compat.numpy`
+(the compatible wrapper for `numpy`) is returned, regardless of the type of the input array.
+This allows us to retain all pre-array API behaviour unless the user has set the environment variable.
+If it isn't set, we raise exceptions for known-bad subclasses
+(as set out in [the RFC](https://github.com/scipy/scipy/issues/18286)),
+like NumPy `MaskedArray`s and object arrays.
+After that, we retrieve the namespace using `array_api_compat.array_namespace`.
+
+The current testing infrastructure allows us to test with PyTorch (CPU & GPU) and CuPy through the command line.
+For example, `python dev.py test -b pytorch` tests with PyTorch, and `python dev.py test -b all` tests with
+NumPy, CuPy, PyTorch and `numpy.array_api`.
+[`numpy.array_api`](https://numpy.org/devdocs/reference/array_api.html) is a strict minimal implementation of the
+standard, which allows us to make sure that our array-agnostic code is _really_ array-agnostic -
+if our tests pass for `numpy.array_api`, then they will pass for any fully compliant library.
+PyTorch GPU can be enabled by setting the environment variable `SCIPY_DEVICE=cuda`.
+
+---
+
+## Status
+
+At the time of writing, SciPy has two of its submodules converted (in the sense I have described above):
+`cluster` and `fft`.
+`linalg` and `signal` have partial coverage in progress, and some of the compiled-code functions in `special`
```

Comment:
And another full stop between module names
![image](https://github.com/Quansight/Quansight-website/assets/51488791/30551639-c5b9-4ac3-a101-8577774bab54)

Reply:
We can update the previous sentence to:

> At the time of writing, SciPy has two of its submodules, `cluster` and `fft`, converted (in the sense I have described above).